### ISSUE TITLE
[Feature] DeepSeek V4.1 Flash M0 剩余问题跟踪：KV 精度、8K 容量、原生接入与 A3/A5 验收

### ISSUE THREAD

**Issue by hashiqiqixian** [hw-native-sys/pypto-serving#240]
### Summary

跟踪 DeepSeek V4.1 Flash 在 Ascend A3/A5 上达到 M0 的剩余问题，重点覆盖 KV cache 精度、8K prefill 容量、原生算子完成度、serving 接入，以及整模型资源和验收证据。优先支持 A3，随后在 A5 上独立验证。

这是跨 `pypto-serving` / `pypto-lib` 的系统级跟踪 issue，后续可以按下文工作项拆分 PR；不是一个 PR 的实现清单。以下事实截至 2026-09-17，来自固定版本源码审计；静态推算、历史测试结果和待实测项分别标注。

**执行路径边界：当前 V4.1 serving 使用自有 Torch-NPU + PyPTO BF16 provider，尚未调用新版库的原生 V4.1 模型目录。因此，原生库中的问题不能直接描述成当前 serving 已发生的运行故障。原生路径完成度与功能 M0 验收需分别记录。**

### Area

Model integration

### Motivation / Use Case

目标是用真实 V4.1 Flash checkpoint，在 8 卡上完成单请求 8192-token prefill、连续 128-token greedy decode，并通过 HTTP 连续请求、请求隔离、缓存释放和失败恢复验证。需要同时明确模型数值、缓存状态和实际资源消耗，避免将单算子测试通过误记为整模型 M0 完成。

### Additional Context

- 审计的 serving 分支：`feat/deepseek-v41-ascend-a3-a5`，固定提交 [`e6455246b8f259e1c487589e69cd106e326b401c`](https://github.com/hashiqiqixian/pypto-serving/commit/e6455246b8f259e1c487589e69cd106e326b401c)，基于 serving main `9dad3ef63ecb0adf3edc38cb4e1585077c8f65ec`。下文 serving 源码链接指向该开发分支的固定提交，不表示相关功能已合入上游 main。
- 实际使用的 pypto-lib 子模块：[`6affa7283a89e83ce9556b6e75449bf56a49944f`](https://github.com/hw-native-sys/pypto-lib/commit/6affa7283a89e83ce9556b6e75449bf56a49944f)。
- 固定参考：`deepseek-ai/DeepSeek-V4.1-Flash@dba1be0a40aa45a94ad051997016db3960a90277`。
- 已有进展：新版库已有 SWA、C2A Full/Reuse、C1A prefill Full/Reindex/Reuse、mHC 和 TP output reduce 的实际实现。
- 已解决项：8K `wo_a` 投影曾估算 824 MiB、超过默认 256 MiB provider 预算；现已按 M 分块修复，重放提交 `2f48db0`。该问题不再作为未解决项。历史 A5 单投影检查通过；rebase 后记录为 746 项 CPU/编译测试通过。这些结果不构成真实权重整模型验收。
- Related: #65（lib/serving 接口职责）、#95（V4 八卡性能观测方法）、#218（SWA 页回收与逻辑页语义）。它们提供相关背景，不代表 V4 的具体故障或性能数值已经在 V4.1 路径复现。

### End-to-End Flow

```text
真实 checkpoint / tokenizer / 固定参考
  -> 权重读取、量化转换、TP/EP 分片与 native packing
  -> 请求编码、页分配、KV/index owner 和 compressor state
  -> 8192-token prefill：attention / mHC / Engram / MoE / LM head
  -> 连续 128-token greedy decode
  -> HTTP 响应、请求结束或失败后的缓存释放/回滚
  -> 独立数值比较、逐卡内存/耗时和可复现 artifacts
```

### Repository / Layer Responsibilities

| 层 | 职责 |
| --- | --- |
| `pypto-lib` | 原生 KV 量化/读写、attention 容量、C1A decode 与 MoE 内核、原生权重/缓存 ABI、算子独立精度验证及平台支持边界 |
| `pypto-serving` | checkpoint 转换和分片、native adapter、页与请求状态生命周期、整模型组合、分布式拓扑、HTTP 及资源观测 |
| 联合验证 | 固定独立 oracle、真实权重数值对齐、A3/A5 各自的 8 卡模型运行、错误恢复和验收 artifacts |

#### 1. KV cache 数值差异（已确认源码差异，原生路径）

- [ ] 修正 FP4 中点舍入并加入独立 byte/scale 回归。库内 [`_nearest_fp4_indices`](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/quantization.py#L69) 使用首项 `argmin`；实际 C1A [`publish_compressed_cache`](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/prefill_c1a_common.py#L314) 和 [`publish_index_cache`](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/prefill_c1a_common.py#L387) 使用 `+0.4999` 后截断。以 group16 `[0.75, 6, 0, ...]`、scale=1 为例，它们将 0.75 舍到 0.5，而固定 ties-even 契约应为 1；首 packed byte 分别为 `0x71` / `0x72`。
- [ ] 将 UE8M0 与 E4M3 scale 的边界分开处理。库内 [`quantize_mxfp4_cache`](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/quantization.py#L120) 对 UE8M0 golden 也使用 `2^-9` floor。group32 `[2^-12, 0, ...]` 因此变为 0；固定参考采用 scale `2^-14`、FP4 code 6，可保留该数值。

上述 scale floor 问题已定位于库内 golden，不外推到所有 publisher；中点问题则有两个真实 C1A publisher 的独立源码证据。当前 serving 未调用这些库函数。验证需覆盖中点两侧、正负零、极小值、饱和边界，以及 payload/scale 的位级一致性，不能仅让 kernel 与共享同一错误的 golden 互相比对。

#### 2. 原生 attention 的单请求 8K 容量（已确认限制）

- [ ] 扩容并验证通信窗口，或实现保持因果性、KV 发布顺序和 compressor state 的内部切分。

[`PREFILL_MAX_TOKENS = 4096`](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/config.py#L345)，[`prefill_tp_output_all_reduce`](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/attention_tp.py#L24) 的窗口据此声明为 4096 行。SWA/C2A harness 拒绝 8192-token prefill；绕过检查直接调用存在静态越界风险，本次未尝试越界执行。

这是与已修复 provider 预算问题独立的限制。DP2 各处理 4096 token 不等于一个 8K 请求；8192 历史 cache 长度测试也不等于 8192 个 query 的 prefill。不得只删除参数检查。

#### 3. 原生算子与模型组合完成度（未实现项）

- [ ] 实现并独立验证 C1A decode 的 [Full](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/decode_c1a_full.py#L158)、[Reindex](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/decode_c1a_reindex.py#L145)、[Reuse](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/decode_c1a_reuse.py#L126)；三个入口目前均抛出 `NotImplementedError`。
- [ ] 实现并验证 [EP MoE](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/moe.py#L132)；现有 ABI/golden 不等于 kernel body 已完成。
- [ ] 明确 Engram 和整模型组合的执行归属：当前原生目录没有 Engram kernel 或完整 40 层模型入口；现有 serving 的对应实现可以作为明确标注的 bridge 路径参与功能验收。

这些是原生路径的交付缺口。若先使用已有 bridge 完成功能 M0，必须记录各模块实际 backend；不得将其标为原生内核已完成。

#### 4. 原生 KV / 权重 / 拓扑接入（已有实现尚未接入）

- [ ] 实现明确的 native adapter，连接 [`backend.py`](https://github.com/hashiqiqixian/pypto-serving/blob/e6455246b8f259e1c487589e69cd106e326b401c/pypto_serving/model/deepseek_v41/backend.py#L323) 当前 provider 路径与新库算子。
- [ ] 对齐缓存布局、128 行物理页、slot/page 映射、因果有效长度、跨层 KV/index owner、ratio-2 pending state 和 commit/rollback。
- [ ] 对齐 checkpoint 权重转换与 native input-major FP8 / `MX_B_NN` scale packing，并独立验证 TP/EP 分片。
- [ ] 明确 8 卡 TP/EP/DP 拓扑和请求/行所有权。库默认 TP4/EP8/DP2，当前 serving 的参考拓扑与之不同，不能仅更新子模块 pin 就认为互通。

serving 的 [`PackedRows`](https://github.com/hashiqiqixian/pypto-serving/blob/e6455246b8f259e1c487589e69cd106e326b401c/pypto_serving/model/deepseek_v41/packed_cache.py#L135) 按行拼接 payload/scale；原生 [`C1A prefill ABI`](https://github.com/hw-native-sys/pypto-lib/blob/6affa7283a89e83ce9556b6e75449bf56a49944f/models/deepseek_v4_1_flash/prefill_c1a_full.py#L214) 使用分离的四维 payload/scale 池。需要真正的适配和生命周期测试，不能只 reshape 或解释同一指针。

#### 5. 整模型激活内存（静态风险，尚未复现 OOM）

- [ ] 观测逐卡峰值，处理 provider 预算以外的完整激活和 workspace；按需要进行分块、缩短生命周期或替换实现。

当前 serving 在 8K 输入下：

| 位置 | 静态尺寸 | 说明 |
| --- | --- | --- |
| [`TensorOps.linear` 的 Engram WKV 输出](https://github.com/hashiqiqixian/pypto-serving/blob/e6455246b8f259e1c487589e69cd106e326b401c/pypto_serving/model/deepseek_v41/numerics.py#L199) | `[8192,25600]` FP32 = 800 MiB/rank | 完整输出；默认 TP 不会将该输出按卡切小 |
| [`hc_post` 广播乘法](https://github.com/hashiqiqixian/pypto-serving/blob/e6455246b8f259e1c487589e69cd106e326b401c/pypto_serving/model/deepseek_v41/numerics.py#L253) | `[8192,4,4,5120]` FP32 = 2.5 GiB/rank | 源码表达式对应的逻辑临时量 |

这些是尺寸推算，不是实测峰值，也不能直接相加当作总 HBM。256 MiB 限制是 provider 局部预算，不是整模型内存上限；现有 memory preflight 也未覆盖全部临时激活。

#### 6. GEMM / 传输 / collective 粒度（静态风险，待 profiling）

- [ ] 在真实维度下记录 kernel launch 数、主机/设备传输量、collective 次数、prefill/decode 耗时和超时；根据结果调整粒度或接入原生算子。

默认 [`matrix_tiles`](https://github.com/hashiqiqixian/pypto-serving/blob/e6455246b8f259e1c487589e69cd106e326b401c/pypto_serving/model/deepseek_v41/tensor_store.py#L240) 采用 N128/K32。单个 Engram WKV `25600×6144` 每层每 rank 需 `(25600/128)×(6144/32)=38,400` 次 GEMM；[`TensorOps.matmul`](https://github.com/hashiqiqixian/pypto-serving/blob/e6455246b8f259e1c487589e69cd106e326b401c/pypto_serving/model/deepseek_v41/numerics.py#L153) 逐次进行 CPU/NPU 往返。8K 时每次 FP32 结果为 4 MiB，累计逻辑结果传输量约 150 GiB/层/rank，不代表常驻内存。

[`indexer`](https://github.com/hashiqiqixian/pypto-serving/blob/e6455246b8f259e1c487589e69cd106e326b401c/pypto_serving/model/deepseek_v41/attention.py#L356) 还逐 query、每 1024 个 index keys 执行 all-reduce。尚未实测完整 8K/8 卡耗时，不能断言必然超时。

#### 7. A3/A5 与真实模型验收（缺少完整证据）

- [ ] A3 优先：明确 bridge/native 支持矩阵，验证所选执行路径、通信与精度；新版原生硬件 harness 主要面向 A5，不能把已有 A3 bridge 结果视为新 native 路径已验证。
- [ ] A5 对所选执行路径独立验证，记录与 A3 的实现差异。
- [ ] 完成真实 checkpoint 的权重转换、单层精度和 8 卡整模型验证；记录单请求 8K prefill 后的 128-token greedy 结果。
- [ ] 完成 HTTP 连续请求、请求隔离、缓存释放/重用、失败回滚与恢复检查，并收集完整 artifacts。

本轮没有重新检查远程权重目录或运行新硬件测试；当前结论是缺少完整验收证据，不是断言远程一定没有权重。已有单投影及 CPU/编译测试不能替代真实权重八卡 HCCL 整模型结果。

### Acceptance Criteria

- [ ] KV 精度：固定独立参考版本；对 SWA、main compressed KV、index KV 记录 payload/scale 与 attention 输出比较，覆盖边界值、跨页、chunk 边界及 Full/Reindex/Reuse 所有权。
- [ ] 8K 容量：同一个请求的 8192 个输入 token 能完成 prefill，随后 decode；窗口访问合法，内部切分（若采用）与参考在因果性和 cache 状态上等价。
- [ ] 执行路径：列出每个模块在 A3/A5 上实际使用的 bridge/native backend、TP/EP/DP 配置和已验证范围。原生 C1A decode/MoE 的完成状态独立记录。
- [ ] 真权重与数值：checkpoint/config/tokenizer/revision 固定；权重转换和单层比较通过；指定文本用例 greedy 生成 128 token 与固定参考对比，报告首个差异及相关 logits，不以同源 golden 自洽替代独立验证。
- [ ] 生命周期：HTTP 连续请求无串扰；完成/失败后释放页面和 pending state；释放后的页面可重用；回滚不会污染后续请求。
- [ ] 资源与可复现性：记录逐卡峰值 HBM、主机内存、prefill/decode 耗时、主要传输/collective 统计；附命令、模型与代码提交、环境版本、日志和 comparator 结果。

### Alternatives Considered

- 保留现有 bridge 先完成可运行的功能 M0，同时按模块接入原生内核；需要诚实记录 backend 和资源代价。
- 8K 可以选择扩容，或在模型内部切分；后者必须验证因果 mask、ratio-2 pending state、跨层共享和 decode 衔接。
- 仅调大 256 MiB provider 预算或更新 pypto-lib 子模块，均不能解决上述跨层问题。

### Out of Scope

- 本 issue 的建立不代表上述功能已实现或已通过实卡验证。
- 不要求把所有工作合并成一个 PR；内核修正、adapter、资源优化和验收可以分别交付。
- 不将全算子融合、全面替换 bridge、吞吐极限优化、MTP、视觉能力或高并发目标自动追加为文本功能 M0 的硬门槛。
- 已修复的 `wo_a` 824 MiB / 256 MiB 问题只保留背景与回归证据，不重复立为未解决缺陷。
